cmd/rotatecerts/wait.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package rotatecerts
+
+import (
+	"time"
+
+	"github.com/Azure/aks-engine/pkg/helpers/ssh"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// WaitForNodesReady returns true if all the cluster nodes reached the Ready state
+func WaitForNodesReady(c Client, nodes []*ssh.RemoteHost, successesNeeded int, interval, timeout time.Duration) error {
+	isNodeReady := func(node v1.Node) bool {
+		for _, c := range node.Status.Conditions {
+			if c.Type == v1.NodeReady {
+				return c.Status == v1.ConditionTrue
+			}
+		}
+		return false
+	}
+
+	successesCount := 0
+	nodeReady := make(map[string]bool, len(nodes))
+	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		nl, err := c.ListNodes()
+		if err != nil {
+			return false, err
+		}
+		for _, n := range nodes {
+			nodeReady[n.URI] = false
+		}
+		for _, nli := range nl.Items {
+			if _, ok := nodeReady[nli.ObjectMeta.Name]; ok {
+				ready := isNodeReady(nli)
+				if !ready {
+					return false, nil
+				}
+				nodeReady[nli.ObjectMeta.Name] = ready
+			}
+		}
+		for _, ready := range nodeReady {
+			if !ready {
+				return false, nil
+			}
+		}
+		successesCount++
+		if successesCount < successesNeeded {
+			return false, nil
+		}
+		return true, nil
+	})
+	return err
+}
+
+// WaitForAllInNamespaceReady returns true if all containers in a given namespace reached the Ready state
```

Comment:
And it looks like these funcs wrap that "let's call both apiservers and assume that the first one we get a response from is the *real* one" thing inside async retry wrapping? So we're basically regularly polling both apiservers until one of them (the valid one) returns a response?

Reply:
yes